### PR TITLE
ENH: Bump for ITK v5.4rc01

### PR DESCRIPTION
**Pull request by tbirdso**
Bump CI and package dependencies for ITK v5.4 release candidate 1.

Expected to include update to C++17 build tools.

### CODE REVIEW

**Review by thewtex — APPROVED** [review]

**Comment by N-Dekker** [comment]
Thanks @tbirdso, really cool! Hope it fixes the CI for PR #236

File: .github/workflows/build-test-package.yml [hunk 1]
```diff
@@ -4,22 +4,22 @@ on: [push,pull_request]
...
     secrets:
       pypi_password: ${{ secrets.pypi_password }}
       
   python-build-workflow-main:
     if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/main' || startsWith(github.ref, 'refs/tags')
-    uses: InsightSoftwareConsortium/ITKRemoteModuleBuildTestPackageAction/.github/workflows/build-test-package-python.yml@edd0a4350396f533e9ec3755ed6c5af5ddfedb97
+    uses: InsightSoftwareConsortium/ITKRemoteModuleBuildTestPackageAction/.github/workflows/build-test-package-python.yml@eebccc8d50a25fac571324ae56dd254bd64607a8
     with:
-      python3-minor-versions: '["7","8","9","10","11"]'
+      python3-minor-versions: '["8","9","10","11"]'
```
Comment (N-Dekker) [inline .github/workflows/build-test-package.yml:22]: I see you removed Python 3.7 support with this commit. Looks fine to me, especially now that Python 3.7 has reached "end of life". According to https://peps.python.org/pep-0537/

> As of 2023-06-27, 3.7 has reached the [end-of-life phase](https://devguide.python.org/devcycle/#end-of-life-branches) of its release cycle.

**Comment by N-Dekker** [comment]
I'm just trying to upgrade ITK from v5.3.0 to v5.4rc01 at the SuperElastix/elastix CI as well: https://github.com/SuperElastix/elastix/tree/Upgrade-ITK-to-v5.4rc01 However, to my surprise, it appears to trigger some regression failures:

From both Windows and Ubuntu, at https://dev.azure.com/kaspermarstal/Elastix/_build/results?buildId=4292&view=results
```
2023-08-26T17:06:52.3783895Z The following tests FAILED:
2023-08-26T17:06:52.3784191Z 	 54 - elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_OVERLAP (Failed)
2023-08-26T17:06:52.3784590Z 	 55 - elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_LANDMARKS (Failed)
2023-08-26T17:06:52.3784910Z 	 83 - elastix_run_3DCT_lung.MI.bspline.SGD.001_COMPARE_TP (Failed)
2023-08-26T17:06:52.3785238Z 	131 - elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads1_COMPARE_TP (Failed)
2023-08-26T17:06:52.3785577Z 	135 - elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads2_COMPARE_TP (Failed)
2023-08-26T17:06:52.3785913Z 	139 - elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads4_COMPARE_TP (Failed)

```

And from macOS, at https://my.cdash.org/viewTest.php?onlyfailed&buildid=2395262

```
6 tests failed.
Name 	Status 	Time 	Details 	Summary
elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_LANDMARKS 	Failed 	390ms 	Completed (Failed) 	Unstable
elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_OVERLAP 	Failed 	6s 870ms 	Completed (Failed) 	Unstable
elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads1_COMPARE_TP 	Failed 	160ms 	Completed (Failed) 	Unstable
elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads2_COMPARE_TP 	Failed 	150ms 	Completed (Failed) 	Unstable
elastix_run_3DCT_lung.MI.bspline.SGD.001-Threads4_COMPARE_TP 	Failed 	150ms 	Completed (Failed) 	Unstable
elastix_run_3DCT_lung.MI.bspline.SGD.001_COMPARE_TP 	Failed 	90ms 	Completed (Failed) 	Unstable

```

Details from https://my.cdash.org/test/90114840 (elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_LANDMARKS):

```
Transforming fixed image landmarks using /Users/runner/work/elastix/elastix/Elastix-build/Testing/elastix_run_3DCT_lung.Kappa.bspline.ASGD.001/TransformParameters.0.txt
Transforming fixed image landmarks using /Users/runner/work/elastix/elastix/Elastix-build/Testing/TransformParameters_3DCT_lung.Kappa.bspline.ASGD.001.txt
The landmark distance between current and baseline is:
min   | Q1    | med   | Q3    | max   | mean
0.000 | 0.000 | 0.504 | 1.135 | 5.791 | 0.755
FAILURE: third quartile landmark distance is higher than 1.0 mm

```
And from https://my.cdash.org/test/90114841 (elastix_run_3DCT_lung.Kappa.bspline.ASGD.001_COMPARE_OVERLAP):

```
Deforming moving image segmentation using /Users/runner/work/elastix/elastix/Elastix-build/Testing/elastix_run_3DCT_lung.Kappa.bspline.ASGD.001/TransformParameters.0.txt
Deforming moving image segmentation using /Users/runner/work/elastix/elastix/Elastix-build/Testing/TransformParameters_3DCT_lung.Kappa.bspline.ASGD.001.txt
The segmentation overlap between current and baseline is 0.973903

FAILURE: overlap is lower than 0.99
```

I do not know if these differences are of any importance, I mean I don't know if these are really buggy, but they certainly suggest that the behavior has changed from ITK v5.3.0 to v5.4rc01. Please let me know if you have a clue about those changes!

**Comment by N-Dekker** [comment]
> I do not know if these differences are of any importance, I mean I don't know if these are really buggy, but they certainly suggest that the behavior has changed from ITK v5.3.0 to v5.4rc01. 

Update: I found that the behavior change between ITK v5.3.0 to v5.4rc01 was caused by pull request https://github.com/InsightSoftwareConsortium/ITK/pull/3929 commit https://github.com/InsightSoftwareConsortium/ITK/commit/4e46cb6ef56658353c14bd58e6b28b63c410f1c8 "STYLE: Default default-constructor of ImageRandom ConstIterator classes", merged on February 24, 2023 (My :anguished: fault!) It should be fixed (restoring the v5.3.0 behavior) by https://github.com/InsightSoftwareConsortium/ITK/pull/4183 "BUG: ImageRandomIteratorWithIndex should not assign data in constructor". I hope this fix can still be included with the next ITK release (candidate).

**Pull request merged**